raft/src/main/java/org/apache/kafka/raft/RequestManager.java

```diff
@@ -17,108 +17,296 @@
+import org.apache.kafka.common.Node;
 
+/**
+ * The request manager keeps tracks of the connection with remote replicas.
+ *
+ * When sending a request update this type by calling {@code onRequestSent(Node, long, long)}. When
+ * the RPC returns a response, update this manager with {@code onResponseResult(Node, long, boolean, long)}.
+ *
+ * Connections start in the ready state ({@code isReady(Node, long)} returns true).
+ *
+ * When a request times out or completes successfully the collection will transition back to the
+ * ready state.
+ *
+ * When a request completes with an error it still transition to the backoff state until
+ * {@code retryBackoffMs}.
+ */
 public class RequestManager {
-    private final Map<Integer, ConnectionState> connections = new HashMap<>();
-    private final List<Integer> voters = new ArrayList<>();
+    private final Map<String, ConnectionState> connections = new HashMap<>();
+    private final ArrayList<Node> bootstrapServers;
 
     private final int retryBackoffMs;
     private final int requestTimeoutMs;
     private final Random random;
 
-    public RequestManager(Set<Integer> voterIds,
-                          int retryBackoffMs,
-                          int requestTimeoutMs,
-                          Random random) {
-
+    public RequestManager(
+        Collection<Node> bootstrapServers,
+        int retryBackoffMs,
+        int requestTimeoutMs,
+        Random random
+    ) {
+        this.bootstrapServers = new ArrayList<>(bootstrapServers);
         this.retryBackoffMs = retryBackoffMs;
         this.requestTimeoutMs = requestTimeoutMs;
-        this.voters.addAll(voterIds);
         this.random = random;
+    }
 
-        for (Integer voterId: voterIds) {
-            ConnectionState connection = new ConnectionState(voterId);
-            connections.put(voterId, connection);
+    /**
+     * Returns true if there any connection with pending requests.
+     *
+     * This is useful for satisfying the invariant that there is only one pending Fetch request.
+     * If there are more than one pending fetch request, it is possible for the follower to write
+     * the same offset twice.
+     *
+     * @param currentTimeMs the current time
+     * @return true if the request manager is tracking at least one request
+     */
+    public boolean hasAnyInflightRequest(long currentTimeMs) {
+        boolean result = false;
+
+        Iterator<ConnectionState> iterator = connections.values().iterator();
+        while (iterator.hasNext()) {
+            ConnectionState connection = iterator.next();
+            if (connection.hasRequestTimedOut(currentTimeMs)) {
+                // Mark the node as ready after request timeout
+                iterator.remove();
+            } else if (connection.isBackoffComplete(currentTimeMs)) {
+                // Mark the node as ready after completed backoff
+                iterator.remove();
+            } else if (connection.hasInflightRequest(currentTimeMs)) {
```

Comment:
It looks like we'll also check hasRequestTimeout in hasInflightRequest. We can consider to do a small optimization here in the future.

Reply:
I agree. I think we can re-implement the RequestManager so that all operations consistently update the state as part of checking the state. I think it would be much easier to read and understand. I opted against it in this PR to minimize the diff and the possibility of introducing a breaking change.